File: DFS:BFS/[DFS:BFS] 영선/단어변환.swift
//
//  단어변환.swift
//  Algorithm
//
//  Created by 임영선 on 2023/10/16.
//

import Foundation

/*
 DFS(문자열, 깊이)
 */

func solution(_ begin:String, _ target:String, _ words:[String]) -> Int {
    var visited = Array(repeating: false, count: words.count)
    var answer = Int.max
    
    func DFS(result: String, depth: Int) {
        if result == target {
            answer = min(answer, depth)
        }
        else if answer > depth {
            for i in 0..<words.count {
                if !visited[i] && isEnabledChange(result, words[i]) {
                    visited[i] = true
                    DFS(result: words[i], depth: depth+1)
                    visited[i] = false
                }
            }
        }
    }
    
    // 1개의 알파벳이 다른 경우만 true 리턴
    func isEnabledChange(_ word1: String, _ word2: String) -> Bool {
        let word1 = word1.map {String($0)}
        let word2 = word2.map {String($0)}
        var countOfDifferent = 0
        for i in 0..<word1.count {
            if word1[i] != word2[i] {
                countOfDifferent += 1
                if countOfDifferent > 1 {
                    return false
                }
            }
        }
        return true
    }
    
    // 타겟이 단어들에 있는 경우만 DFS 돌림
    if words.contains(target) {
        for i in 0..<words.count {
            // 바꿀 수 있는 문자열들만 DFS돌림
            if isEnabledChange(words[i], begin) {
                visited = Array(repeating: false, count: words.count)
                DFS(result: words[i], depth: 1)
            }
        }
    } else {
        return 0
    }
    
    return answer
    
}

print(solution("hit", "cog", ["hot", "dot", "dog", "lot", "log", "cog"]))

File: DFS_BFS/[지후] DFS_BFS/단어변환.js
function solution(begin, target, words) {
    // target이 words 안에 없을 시 0 return
    if (!words.includes(target)) {
        return 0
    }

    const visited = { [begin]: 0 };
    const queue = [begin];

    // BFS
    while (queue.length) {
        const current = queue.shift();

        // queue 에서 지운 값이 target과 일치한 경우 종료
        if (current === target) break;

        words.forEach(value => {
            if (checking(value, current) && !visited[value]) {
                visited[value] = visited[current] + 1;
                queue.push(value);
            }
        })
    }

    const result = visited[target] ?? 0;

    return result;
}

const checking = (str1, str2) => {
    let count = 0;
    const len = str1.length;

    for (let i = 0; i < len; i++) {
        if (str1[i] !== str2[i]) count++;
    }


    // 서로 다른 철자가 딱 1개면 true
    return count === 1 ? true : false;
}

solution("hit", "cog", ["hot", "dot", "dog", "lot", "log", "cog"]);